File: gin/internal/model/players/player.go
package model_players

import (
	"time"

	"gorm.io/gorm"
	"fmt"
)

type PlayerRankType string
const (
	PlayerRankType_Bronze PlayerRankType = "Bronze"
	PlayerRankType_Silver PlayerRankType = "Silver"
	PlayerRankType_Gold PlayerRankType = "Gold"
	PlayerRankType_Platinum PlayerRankType = "Platinum"
	PlayerRankType_Diamond PlayerRankType = "Diamond"
	PlayerRankType_Master PlayerRankType = "Master"
	PlayerRankType_Grandmaster PlayerRankType = "Grandmaster"
)

type PlayerPreferredFormatType string
const (
	PlayerPreferredFormatType_Standard PlayerPreferredFormatType = "Standard"
	PlayerPreferredFormatType_Extended PlayerPreferredFormatType = "Extended"
	PlayerPreferredFormatType_Legacy PlayerPreferredFormatType = "Legacy"
	PlayerPreferredFormatType_Vintage PlayerPreferredFormatType = "Vintage"
	PlayerPreferredFormatType_Commander PlayerPreferredFormatType = "Commander"
	PlayerPreferredFormatType_Draft PlayerPreferredFormatType = "Draft"
)

// PlayerCreateRequest is the POST body.
type PlayerCreateRequest struct {
	DisplayName string `json:"display_name" binding:"required"`
	Rank PlayerRankType `json:"rank" binding:"required"`
	Rating int `json:"rating"`
	PeakRating int `json:"peak_rating"`
	Bio *string `json:"bio"`
	CountryCode *string `json:"country_code"`
	AvatarUrl *string `json:"avatar_url"`
	PreferredFormat *PlayerPreferredFormatType `json:"preferred_format"`
	IsVerified bool `json:"is_verified"`
	LastActiveAt *string `json:"last_active_at"`
	UserID uint `json:"user_id"`
}

// PlayerUpdateRequest is the PUT/PATCH body — all fields optional.
type PlayerUpdateRequest struct {
	DisplayName *string `json:"display_name"`
	Rank *PlayerRankType `json:"rank"`
	Rating *int `json:"rating"`
	PeakRating *int `json:"peak_rating"`
	Bio *string `json:"bio"`
	CountryCode *string `json:"country_code"`
	AvatarUrl *string `json:"avatar_url"`
	PreferredFormat *PlayerPreferredFormatType `json:"preferred_format"`
	IsVerified *bool `json:"is_verified"`
	LastActiveAt *string `json:"last_active_at"`
	UserID *uint `json:"user_id"`
}

// PlayerResponse is the JSON representation returned by the API.
type PlayerResponse struct {
	ID        uint      `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	DisplayName string `json:"display_name"`
	Rank PlayerRankType `json:"rank"`
	Rating int `json:"rating"`
	PeakRating int `json:"peak_rating"`
	Bio *string `json:"bio"`
	CountryCode *string `json:"country_code"`
	AvatarUrl *string `json:"avatar_url"`
	PreferredFormat *PlayerPreferredFormatType `json:"preferred_format"`
	IsVerified bool `json:"is_verified"`
	LastActiveAt *string `json:"last_active_at"`
	UserID uint `json:"user_id"`
}

type Player struct {
	gorm.Model
	DisplayName string `gorm:"column:display_name;not null"`
	Rank PlayerRankType `gorm:"column:rank;not null;default:'Bronze'"`
	Rating int `gorm:"column:rating;not null;default:1000"`
	PeakRating int `gorm:"column:peak_rating;not null;default:1000"`
	Bio *string `gorm:"column:bio;type:text"`
	CountryCode *string `gorm:"column:country_code"`
	AvatarUrl *string `gorm:"column:avatar_url"`
	PreferredFormat *PlayerPreferredFormatType `gorm:"column:preferred_format"`
	IsVerified bool `gorm:"column:is_verified;default:false"`
	LastActiveAt *string `gorm:"column:last_active_at"`
	UserID uint `gorm:"column:user_id"`
	AchievementsIDs []uint `gorm:"serializer:json;column:achievements_ids"`
	FriendsIDs []uint `gorm:"serializer:json;column:friends_ids"`
}

func (m *Player) ToResponse() PlayerResponse {
	return PlayerResponse{
		ID:        m.ID,
		CreatedAt: m.CreatedAt,
		UpdatedAt: m.UpdatedAt,
		DisplayName: m.DisplayName,
		Rank: m.Rank,
		Rating: m.Rating,
		PeakRating: m.PeakRating,
		Bio: m.Bio,
		CountryCode: m.CountryCode,
		AvatarUrl: m.AvatarUrl,
		PreferredFormat: m.PreferredFormat,
		IsVerified: m.IsVerified,
		LastActiveAt: m.LastActiveAt,
		UserID: m.UserID,
	}
}

func (m *Player) ApplyUpdate(req PlayerUpdateRequest) {
	if req.DisplayName != nil { m.DisplayName = *req.DisplayName }
	if req.Rank != nil { m.Rank = *req.Rank }
	if req.Rating != nil { m.Rating = *req.Rating }
	if req.PeakRating != nil { m.PeakRating = *req.PeakRating }
	if req.Bio != nil { m.Bio = req.Bio }
	if req.CountryCode != nil { m.CountryCode = req.CountryCode }
	if req.AvatarUrl != nil { m.AvatarUrl = req.AvatarUrl }
	if req.PreferredFormat != nil { m.PreferredFormat = req.PreferredFormat }
	if req.IsVerified != nil { m.IsVerified = *req.IsVerified }
	if req.LastActiveAt != nil { m.LastActiveAt = req.LastActiveAt }
	if req.UserID != nil { m.UserID = *req.UserID }
}

func (m *Player) Promote()  (bool, error) {
	return false, fmt.Errorf("Promote: not implemented")
}

func (m *Player) Demote()  (bool, error) {
	return false, fmt.Errorf("Demote: not implemented")
}

func (m *Player) RecordWin()  error {
	return fmt.Errorf("RecordWin: not implemented")
}

func (m *Player) RecordLoss()  error {
	return fmt.Errorf("RecordLoss: not implemented")
}

func (m *Player) WinRate()  (float64, error) {
	return 0.0, fmt.Errorf("WinRate: not implemented")
}

func (m *Player) Verify()  error {
	return fmt.Errorf("Verify: not implemented")
}

func (m *Player) UpdateRating(delta int)  error {
	return fmt.Errorf("UpdateRating: not implemented")
}

File: axum/src/models/players/player.rs
#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum PlayerRank {
    Bronze,
    Silver,
    Gold,
    Platinum,
    Diamond,
    Master,
    Grandmaster,
}

impl PlayerRank {
    pub fn as_str(&self) -> &'static str {
        match self {
            PlayerRank::Bronze => "Bronze",
            PlayerRank::Silver => "Silver",
            PlayerRank::Gold => "Gold",
            PlayerRank::Platinum => "Platinum",
            PlayerRank::Diamond => "Diamond",
            PlayerRank::Master => "Master",
            PlayerRank::Grandmaster => "Grandmaster",
        }
    }
}

impl std::fmt::Display for PlayerRank {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for PlayerRank {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Bronze" => Ok(PlayerRank::Bronze),
            "Silver" => Ok(PlayerRank::Silver),
            "Gold" => Ok(PlayerRank::Gold),
            "Platinum" => Ok(PlayerRank::Platinum),
            "Diamond" => Ok(PlayerRank::Diamond),
            "Master" => Ok(PlayerRank::Master),
            "Grandmaster" => Ok(PlayerRank::Grandmaster),
            _ => Err(format!("unknown PlayerRank: {}", s)),
        }
    }
}

impl From<String> for PlayerRank {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid PlayerRank: {}", s))
    }
}

#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum PlayerPreferredFormat {
    Standard,
    Extended,
    Legacy,
    Vintage,
    Commander,
    Draft,
}

impl PlayerPreferredFormat {
    pub fn as_str(&self) -> &'static str {
        match self {
            PlayerPreferredFormat::Standard => "Standard",
            PlayerPreferredFormat::Extended => "Extended",
            PlayerPreferredFormat::Legacy => "Legacy",
            PlayerPreferredFormat::Vintage => "Vintage",
            PlayerPreferredFormat::Commander => "Commander",
            PlayerPreferredFormat::Draft => "Draft",
        }
    }
}

impl std::fmt::Display for PlayerPreferredFormat {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for PlayerPreferredFormat {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Standard" => Ok(PlayerPreferredFormat::Standard),
            "Extended" => Ok(PlayerPreferredFormat::Extended),
            "Legacy" => Ok(PlayerPreferredFormat::Legacy),
            "Vintage" => Ok(PlayerPreferredFormat::Vintage),
            "Commander" => Ok(PlayerPreferredFormat::Commander),
            "Draft" => Ok(PlayerPreferredFormat::Draft),
            _ => Err(format!("unknown PlayerPreferredFormat: {}", s)),
        }
    }
}

impl From<String> for PlayerPreferredFormat {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid PlayerPreferredFormat: {}", s))
    }
}

#[derive(Debug, Clone, sqlx::FromRow, serde::Serialize)]
pub struct Player {
    pub id: i64,
    pub created_at: String,
    pub updated_at: String,
    pub display_name: String,
    pub rank: PlayerRank,
    pub rating: i64,
    pub peak_rating: i64,
    pub bio: Option<String>,
    pub country_code: Option<String>,
    pub avatar_url: Option<String>,
    pub preferred_format: Option<PlayerPreferredFormat>,
    #[serde(serialize_with = "crate::serde_utils::serialize_i64_as_bool")]
    pub is_verified: i64,
    #[serde(rename = "lastActiveAt")]
    pub last_active_at: Option<String>,
    pub user_id: Option<i64>,
}

#[derive(Debug, serde::Deserialize)]
pub struct PlayerCreateRequest {
    pub display_name: String,
    pub rank: PlayerRank,
    pub rating: i64,
    pub peak_rating: i64,
    pub bio: Option<String>,
    pub country_code: Option<String>,
    pub avatar_url: Option<String>,
    pub preferred_format: Option<PlayerPreferredFormat>,
    pub is_verified: bool,
    pub last_active_at: Option<String>,
    pub user_id: Option<i64>,
}

#[derive(Debug, serde::Deserialize)]
pub struct PlayerUpdateRequest {
    pub display_name: Option<String>,
    pub rank: Option<PlayerRank>,
    pub rating: Option<i64>,
    pub peak_rating: Option<i64>,
    pub bio: Option<String>,
    pub country_code: Option<String>,
    pub avatar_url: Option<String>,
    pub preferred_format: Option<PlayerPreferredFormat>,
    pub is_verified: Option<bool>,
    pub last_active_at: Option<String>,
    pub user_id: Option<i64>,
}
